src/core/atomic.d

```diff
@@ -1404,6 +1410,328 @@ else version (AsmX86_64)
+            static if (T.sizeof == ubyte.sizeof)
+            {
+                __atomic_store_1(&val, *cast(ubyte*) &newval, ms);
+            }
+            else static if (T.sizeof == ushort.sizeof)
+            {
+                __atomic_store_2(&val, *cast(ushort*) &newval, ms);
+            }
+            else static if (T.sizeof == uint.sizeof)
+            {
+                __atomic_store_4(&val, *cast(uint*) &newval, ms);
+            }
+            else static if (T.sizeof == ulong.sizeof && GNU_Have_64Bit_Atomics)
+            {
+                __atomic_store_8(&val, *cast(ulong*) &newval, ms);
+            }
+            else static if (GNU_Have_LibAtomic)
+            {
+                __atomic_store(T.sizeof, &val, cast(void*)&newval, ms);
+            }
+            else
+                static assert(0, "Invalid template type specified.");
+        }
+        else
+        {
+            getAtomicMutex.lock();
+            val = newval;
+            getAtomicMutex.unlock();
+        }
+    }
+
+    void atomicFence() nothrow @nogc
+    {
+        static if (GNU_Have_Atomics || GNU_Have_LibAtomic)
+            __atomic_thread_fence(MemoryOrder.seq);
+        else
+        {
+            getAtomicMutex.lock();
+            getAtomicMutex.unlock();
+        }
+    }
+
+    static if (!GNU_Have_Atomics && !GNU_Have_LibAtomic)
+    {
+        // Use system mutex for atomics, faking the purity of the functions so
+        // that they can be used in pure/nothrow/@safe code.
+        extern (C) private pure @trusted @nogc nothrow
+        {
+            static if (GNU_Thread_Model == ThreadModel.Posix)
+            {
+                import core.sys.posix.pthread;
+                alias atomicMutexHandle = pthread_mutex_t;
+
+                pragma(mangle, "pthread_mutex_init") int fakePureMutexInit(pthread_mutex_t*, pthread_mutexattr_t*);
+                pragma(mangle, "pthread_mutex_lock") int fakePureMutexLock(pthread_mutex_t*);
+                pragma(mangle, "pthread_mutex_unlock") int fakePureMutexUnlock(pthread_mutex_t*);
+            }
+            else static if (GNU_Thread_Model == ThreadModel.Win32)
+            {
+                import core.sys.windows.winbase;
+                alias atomicMutexHandle = CRITICAL_SECTION;
+
+                pragma(mangle, "InitializeCriticalSection") int fakePureMutexInit(CRITICAL_SECTION*);
+                pragma(mangle, "EnterCriticalSection") void fakePureMutexLock(CRITICAL_SECTION*);
+                pragma(mangle, "LeaveCriticalSection") int fakePureMutexUnlock(CRITICAL_SECTION*);
+            }
+            else
+            {
+                alias atomicMutexHandle = int;
+            }
+        }
+
+        // Implements lock/unlock operations.
+        private struct AtomicMutex
+        {
+            int lock() pure @trusted @nogc nothrow
+            {
+                static if (GNU_Thread_Model == ThreadModel.Posix)
+                {
+                    if (!_inited)
```

Comment:
Not sure if it is critical but this looks a bit dangerous if initialization is requested from two threads concurrently.
What platforms match the condition (!GNU_Have_Atomics && !GNU_Have_LibAtomic)? I guess Windows will always have atomics.

Reply:
I think the only way that could happen is if a C program calls rt_init in two threads concurrently.  Using pragma(crt_constructor) is another option, though I don't recall adding that feature to the dmd-cxx branch.

Reply:
Because there is an atomic operation somewhere called from rt_init? A comment to that respect would be nice in that case.

Reply:
Hmm, I did write this at the 11th hour, so wasn't strictly thinking too clear about this.  Also this was committed before @jpf91 added gcc.gthreads, so I guess we could remove some duplication here.

Reply:
I think something like this would be more succinct, and doesn't have the same pitfall.
```
static if (!GNU_Have_Atomics && !GNU_Have_LibAtomic)
{
    import gcc.gthread;

    // Internal static mutex reference.
    private __gthread_mutex_t* _getAtomicMutex() @trusted @nogc nothrow @property
    {
        __gshared static __gthread_mutex_t mutex = GTHREAD_MUTEX_INIT;
        return &mutex;
    }

    // Implements lock/unlock operations.
    private int _atomicMutexLock(__gthread_mutex_t* mutex) @trusted @nogc nothrow
    {
        return __gthread_mutex_lock(mutex);
    }

    private int _atomicMutexUnlock(__gthread_mutex_t* mutex) @trusted @nogc nothrow
    {
        return __gthread_mutex_unlock(mutex);
    }

    // Fake the purity of the functions so that they can be used in pure/nothrow/@safe code.
    pragma(mangle, _getAtomicMutex.mangleof)
    private __gthread_mutex_t* getAtomicMutex() pure @trusted @nogc nothrow @property;

    pragma(mangle, _atomicMutexLock.mangleof)
    private int atomicMutexLock(__gthread_mutex_t*) pure @trusted @nogc nothrow;

    pragma(mangle, _atomicMutexUnlock.mangleof)
    private int atomicMutexUnlock(__gthread_mutex_t*) pure @trusted @nogc nothrow;
}
```
Once __gthread_mutex_t and functions have been added to gcc.gthread, that is.

Reply:
This should work. However, I don't really like that we duplicate `core.sync.mutex` here. Is this only required because `core.sync.mutex` is not `pure`? If so, can we change it to be pure?